### PR TITLE
virt-launcher: drop CAP_NET_RAW from compute container

### PR DESCRIPTION
**Pull request by AlonaKaplan**
Binding a socket to the pod bridge using SO_BINDTODEVICE requires
CAP_NET_RAW.
However, since the DHCP request is a broadcast message, the dns server is listening on 0.0.0.0.
It means that without SO_BINDTODEVICE  to the bridge we have 2 problems.
1. We may get dhcp requests on any interface of the pod.
2. Even when we get DHCP request from the vm (on the brdige) the response may be sent from another interface of the pod and never get to the vm.

To workaround the issues we set `ipv4.FlagInterface` on the connection (it means the socket will pass [IP_PKT‐INFO.ipi_ifindex](https://linux.die.net/man/7/ip) with the interface index).
1. We filter the received packets by the bridge interface name.
2. Since the `ipi_ifindex` is set, the bridge ip is used as the local source address for the routing table lookup and for setting up IP source route options.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
According to the documentation, SO_BROADCAST has to be set since the dhcp OFFER is sent over broadcast. However, it succeeds to work without SO_BROADCAST.
SO_BROADCAST doesn't require net_admin or net_raw, so we may add it in the future if we find out that it is needed.

**Release note**:

```release-note
CAP_NET_RAW removed from virt-launcher.
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/network/dhcp/filter.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		Control: func(network, address string, c syscall.RawConn) error {
+			var opErr error
+			err := c.Control(func(fd uintptr) {
+				opErr = syscall.SetsockoptInt(int(fd), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 1)
+			})
+			if err != nil {
+				return err
+			}
+			return opErr
+		},
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/dhcp/filter.go:29]: Can you move this to a function with a proper name? Something like `setSOReuseaddr`.

File: pkg/virt-launcher/virtwrap/network/dhcp/filter.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+type serveIfConn struct {
+	ifIndex int
+	conn    *ipv4.PacketConn
+	cm      *ipv4.ControlMessage
+}
+
+func (s *serveIfConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
+	for { // Filter all other interfaces
+		n, s.cm, addr, err = s.conn.ReadFrom(b)
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/dhcp/filter.go:56]: I find it "dangerous" to assign `s.cm` before assuring the read succeeded.
It changes the object state as a side effect and the caller may still perform a `WriteTo` even though the value is not populated correctly (e.g. it matched `s.cm == nil`).

In the condition below, if the `s.cm` has no value, it returns without an error and without a proper filtering.

File: pkg/virt-controller/services/template.go [hunk 3]
```diff
@@ -848,7 +848,8 @@ func (t *templateService) RenderLaunchManifest(vmi *v1.VirtualMachineInstance) (
 			RunAsUser:  &userId,
 			Privileged: &privileged,
 			Capabilities: &k8sv1.Capabilities{
-				Add: capabilities,
+				Add:  capabilities,
+				Drop: []k8sv1.Capability{CAP_NET_RAW},
```
Comment (EdDev) [inline pkg/virt-controller/services/template.go:852]: Can you please specify why an explicit drop is needed in the commit message? Other capabilities are just not added, but I guess this one is enabled by default with some CRI implementations.

**Review by EdDev — CHANGES_REQUESTED** [review]
Nice one.
Except the inline comments, I think the new `filter` file may fit a different name. It's core functionality is to implement the `server`, the filtering part seems like a small part of it.

File: pkg/virt-launcher/virtwrap/network/dhcp/dhcp.go [hunk 4]
```diff
@@ -87,12 +86,13 @@ func SingleClientDHCPServer(
...
 	}
 
-	l, err := dhcpConn.NewUDP4BoundListener(serverIface, ":67")
+	l, err := NewUDP4FilterListener(serverIface, ":67")
 	if err != nil {
 		return err
 	}
 	defer l.Close()
 	err = dhcp.Serve(l, handler)
+
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/dhcp/dhcp.go:95]: What change? The new line?

[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/virt-launcher/virtwrap/network/dhcp/dhcp.go:95]: Yes.

[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/dhcp/dhcp.go:95]: done

File: pkg/virt-launcher/virtwrap/network/dhcp/filter.go (hunk 1 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/dhcp/filter.go:29]: Extracted to `CreateListenConfig`.

File: pkg/virt-launcher/virtwrap/network/dhcp/filter.go (hunk 2 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/dhcp/filter.go:56]: Changed to use `dhcpConn.NewServeIf`, so now I cannot change the method.

File: pkg/virt-launcher/virtwrap/network/dhcp/filter.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// This file is inspired by github.com/krolaw/dhcp4/conn/filter.go.
+// Couldn't use the original one since SO_REUSEADDR option has to be set.
```
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/dhcp/filter.go:2]: Changed the code to use `dhcpConn.NewServeIf`, so now the is no code copied from github.com/krolaw/dhcp4/conn/filter.go.

File: pkg/virt-controller/services/template.go (hunk 3 above)
Comment (AlonaKaplan) [reply pkg/virt-controller/services/template.go:857]: yes, this one may be enabled by default. Done.

File: pkg/virt-launcher/virtwrap/network/dhcp/BUILD.bazel [hunk 6]
```diff
@@ -5,6 +5,7 @@ go_library(
     srcs = [
         "dhcp.go",
         "ethtool.go",
+        "socket_listener.go",
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/dhcp/BUILD.bazel:8]: nit: I prefer without the `_`, but there is no consensus in general.
https://github.com/golang/go/issues/36060

File: pkg/virt-launcher/virtwrap/network/dhcp/socket_listener.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+			})
+			if err != nil {
+				return err
+			}
+			return opErr
+		},
+	}
+}
+
+type ServeIfConn interface {
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/network/dhcp/socket_listener.go:70]: nit: This should have appeared at the head of the file IMO.

**Review by EdDev — APPROVED** [review]
Thank you, looks nice.
I had only two nits, none of them a must for me.

File: pkg/virt-launcher/virtwrap/network/dhcp/BUILD.bazel (hunk 6 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/network/dhcp/BUILD.bazel:8]: done

**Pull request merged**